### ISSUE TITLE
Consider adding support in WebIDL for defining template tags

### ISSUE THREAD

**Issue by littledan** [heycam/webidl#631]
The ES6 feature of [tagged template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) has recently been catching on in JavaScript. We may want to add Web APIs which define template tags in the near future, e.g., [for  Trusted Types, to test literal-ness](https://github.com/WICG/trusted-types/issues/96) (cc @koto) or for a built-in version of something like the parsing part of [lit-html](https://github.com/polymer/lit-html) (cc @justinfagnani).

Would it make sense to add support to WebIDL to add template tags? What would that look like?

**Comment by saschanaz** [comment]
Something like `template identifier = return_type (/* arguments... */);` where arguments correspond to expressions?

**Comment by saschanaz** [comment]
AFAIK tagged template functions are just functions, so we may just need a way to define general global (or module level) functions.

**Comment by littledan** [comment]
You can make global functions by defining operations in a global interface. For modules, see https://github.com/heycam/webidl/issues/592 .

**Comment by domenic** [comment]
Is there something preventing you from just using the normal operations syntax for template tags?

**Comment by littledan** [comment]
@domenic Well, you'd need to take it in as an `object` and use a bunch of ECMAScript operations to get the template object, right?

**Comment by domenic** [comment]
Wouldn't just `ReturnType someTag(sequence<DOMString> parts, TypesInteropolated interpolated...)` work? Then you'd get two Infra lists, _parts_ and _interopolated_, each of which have the correct Web IDL type.

**Comment by littledan** [comment]
Yeah, that could work for some things. Some additional things that would be nice to do in a template tag:
- associate something with the template object (e.g., to reduce overhead the second time it's called; maybe this would typically be an unobservable optimization though)
- use the "raw" Array
- check whether the object passed in is "really" a template object

**Comment by justinfagnani** [comment]
Slightly off-topic for webidl, but since you tagged me: I think the most interesting future direction is for JS to allow host-defined tags that are _not identifiers_ so that the host can switch parsers for the template literal contents before evaluating any JS.

We're running out of syntactic space, but imagine that:

```js
let title = (text) => #html`<h1>${text}</h1>`;
```

invoked the host-defined parser for `#html` at parse time, and at eval time invoked the associated host-defined tag function, but instead of a template object as the first argument, the first argument is an object produced by the host-defined parser.

**Comment by littledan** [comment]
Possible place to follow up on this idea: https://github.com/littledan/proposal-reserved-decorator-like-syntax
